librubyfmt/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+#![deny(warnings, missing_copy_implementations)]
+use std::ffi::CString;
+#[macro_use]
+extern crate lazy_static;
+
+use std::io::{BufReader, Cursor, Write};
+use std::slice;
+
+#[global_allocator]
+static ALLOC: jemallocator::Jemalloc = jemallocator::Jemalloc;
+
+pub type RawStatus = i64;
+
+mod breakable_entry;
+mod comment_block;
+mod de;
+mod delimiters;
+mod file_comments;
+mod format;
+mod intermediary;
+mod line_metadata;
+mod line_tokens;
+mod parser_state;
+mod render_queue_writer;
+mod ripper_tree_types;
+mod ruby;
+mod types;
+
+use file_comments::FileComments;
+use parser_state::ParserState;
+use ruby::VALUE;
+
+#[cfg(debug_assertions)]
+use log::debug;
+#[cfg(debug_assertions)]
+use simplelog::{Config, LevelFilter, TermLogger, TerminalMode};
+
+type RubyfmtResult<T = ()> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+extern "C" {
+    pub fn Init_ripper();
+}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct FormatBuffer {
+    pub bytes: *const libc::c_char,
+    pub count: i64,
+}
+
+impl FormatBuffer {
+    pub fn into_buf(self) -> &'static [u8] {
+        unsafe { slice::from_raw_parts(self.bytes as *const u8, self.count as usize) }
+    }
+
+    pub fn into_string(self) -> String {
+        unsafe {
+            let vec = Vec::from_raw_parts(
+                self.bytes as *mut u8,
+                self.count as usize,
+                self.count as usize,
+            );
+            String::from_utf8_unchecked(vec)
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum InitStatus {
+    OK = 0,
+    ERROR = 1,
+}
+
+#[no_mangle]
+pub extern "C" fn rubyfmt_init() -> libc::c_int {
+    init_logger();
+    unsafe {
+        ruby::ruby_init();
+    }
+    let res = load_ripper();
+    if res.is_err() {
+        return InitStatus::ERROR as libc::c_int;
+    }
+
+    let res = load_rubyfmt();
+    if res.is_err() {
+        return InitStatus::ERROR as libc::c_int;
+    }
+
+    InitStatus::OK as libc::c_int
+}
+
+pub fn format_buffer(buf: String) -> String {
+    eprintln!("format 1");
+    let bytes: Vec<libc::c_char> = buf
+        .into_bytes()
+        .into_iter()
+        .map(|v| v as libc::c_char)
+        .collect();
```

Comment:
This is moot since you're getting rid of `FormatBuffer`, but you could just cast the pointer for the same effect but without a copy